### PR TITLE
Architecture: Rename ClientApi -> SystemApi + Move CryptoUtils into VM Layer

### PR DESCRIPTION
**Pull request by talekhinezh**
* Rename ClientApi -> SystemApi
* Move CryptoUtils from System Layer into VM Layer

### CODE REVIEW

File: radix-engine/src/system/system.rs [hunk 1]
```diff
@@ -2869,99 +2884,7 @@ where
...
-                    sizes: sizes.as_slice(),
-                },
-            )?;
-            Ok(aggregate_verify_bls12381_v1(pub_keys_and_msgs, signature) as u32)
-        } else {
-            Err(RuntimeError::SystemError(SystemError::InputDataEmpty))
-        }
-    }
-
-    #[trace_resources(log=message.len(), log=public_keys.len())]
```
Comment (mstrug-rdx) [inline radix-engine/src/system/system.rs:2912]: `trace_resource` macro should be moved to `fast_aggregate_verify_bls12381_v1()` function.

File: radix-engine/src/system/system.rs [hunk 2]
```diff
@@ -2869,99 +2884,7 @@ where
...
-        self.api.kernel_get_system().modules.apply_execution_cost(
-            ExecutionCostingEntry::Bls12381V1Verify {
-                size: message.len(),
-            },
-        )?;
-        Ok(verify_bls12381_v1(message, public_key, signature) as u32)
-    }
-
-    // Trace average message length and number of public_keys
-    #[trace_resources(log={pub_keys_and_msgs.iter().flat_map(|(_, msg)| msg).count()/pub_keys_and_msgs.len()},log=pub_keys_and_msgs.len())]
```
Comment (mstrug-rdx) [inline radix-engine/src/system/system.rs:2893]: `trace_resource` macro should be moved to `aggregate_verify_bls12381_v1()` function.

File: radix-engine/src/system/system.rs [hunk 3]
```diff
@@ -2869,99 +2884,7 @@ where
     feature = "std",
     catch_unwind(crate::utils::catch_unwind_system_panic_transformer)
 )]
-impl<'a, Y, V> ClientCryptoUtilsApi<RuntimeError> for SystemService<'a, Y, V>
-where
-    Y: KernelApi<System<V>>,
-    V: SystemCallbackObject,
-{
-    #[trace_resources(log=message.len())]
```
Comment (mstrug-rdx) [inline radix-engine/src/system/system.rs:2877]: `trace_resource` macro should be moved to `verify_bls12381_v1()` function.

File: radix-engine/src/system/system.rs [hunk 4]
```diff
@@ -2869,99 +2884,7 @@ where
...
-                    keys_cnt: public_keys.len(),
-                },
-            )?;
-            Ok(fast_aggregate_verify_bls12381_v1(message, public_keys, signature) as u32)
-        } else {
-            Err(RuntimeError::SystemError(SystemError::InputDataEmpty))
-        }
-    }
-
-    #[trace_resources(log=signatures.len())]
```
Comment (mstrug-rdx) [inline radix-engine/src/system/system.rs:2932]: `trace_resource` macro should be moved to `Bls12381G2Signature::aggregate()` function.

File: radix-engine/src/system/system.rs [hunk 5]
```diff
@@ -2869,99 +2884,7 @@ where
...
-                },
-            )?;
-            Bls12381G2Signature::aggregate(signatures)
-                .map_err(|err| RuntimeError::SystemError(SystemError::BlsError(err.to_string())))
-        } else {
-            Err(RuntimeError::SystemError(SystemError::InputDataEmpty))
-        }
-    }
-
-    #[trace_resources(log=data.len())]
```
Comment (mstrug-rdx) [inline radix-engine/src/system/system.rs:2950]: `trace_resource` macro should be moved to `keccak256_hash()` function.

File: radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs [hunk 6]
```diff
@@ -588,10 +591,18 @@ where
             scrypto_decode(&pub_keys_and_msgs)
                 .map_err(WasmRuntimeError::InvalidBlsPublicKeyOrMessage)?;
 
-        let result = self
-            .api
-            .bls12381_v1_aggregate_verify(&pub_keys_and_msgs, &signature)?;
-        Ok(result)
+        if pub_keys_and_msgs.is_empty() {
+            return Err(InvokeError::SelfError(WasmRuntimeError::InputDataEmpty));
+        }
```
Comment (iamyulong) [inline radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs:596]: Is this a breaking change?

File: radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs [hunk 7]
```diff
@@ -605,10 +616,17 @@ where
         let signature: Bls12381G2Signature =
             scrypto_decode(&signature).map_err(WasmRuntimeError::InvalidBlsSignature)?;
 
-        let result =
-            self.api
-                .bls12381_v1_fast_aggregate_verify(&message, &public_keys, &signature)?;
-        Ok(result)
+        if public_keys.is_empty() {
+            return Err(InvokeError::SelfError(WasmRuntimeError::InputDataEmpty));
+        }
```
Comment (iamyulong) [inline radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs:621]: Ditto

File: radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs [hunk 8]
```diff
@@ -618,7 +636,16 @@ where
         let signatures: Vec<Bls12381G2Signature> =
             scrypto_decode(&signatures).map_err(WasmRuntimeError::InvalidBlsSignature)?;
 
-        let agg_sig = self.api.bls12381_g2_signature_aggregate(&signatures)?;
+        if signatures.is_empty() {
+            return Err(InvokeError::SelfError(WasmRuntimeError::InputDataEmpty));
+        }
```
Comment (iamyulong) [inline radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs:641]: Ditto

**Review by mstrug-rdx — APPROVED** [review]
When removing crypto bls related functions, you also removed `trace_resource` macro which was used to calculate costing coefficients. I've added comments to that macros to move them to lower layer function in case we need to run costing evaluation again.

Also I saw in couple places that `KernelNodeApi + ClientApi` was changed to `SystemApi`, this is probably intentional but just wanted to point that out.

File: radix-engine/src/system/system.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/system/system.rs:2912]: good point!

File: radix-engine/src/system/system.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine/src/system/system.rs:2893]: good point!

File: radix-engine/src/system/system.rs (hunk 3 above)
Comment (talekhinezh) [reply radix-engine/src/system/system.rs:2877]: good point!

File: radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs (hunk 6 above)
Comment (talekhinezh) [reply radix-engine/src/vm/wasm_runtime/scrypto_runtime.rs:599]: The check was in the `api.bls12381_v1_aggregate_verify()` call, the logic of which has now moved into this function.

**Review by iamyulong — APPROVED** [review]

**Pull request merged**